File: 2025/02 February/20 Find median in a stream/main.py
# Link: https://www.geeksforgeeks.org/problems/find-median-in-a-stream-1587115620/1

import heapq


class Solution:
    def getMedian(self, arr):
        h1, h2, r = [], [], []
        for i in arr:
            heapq.heappush(h1, -heapq.heappushpop(h2, i))
            if len(h1) > len(h2):
                heapq.heappush(h2, -heapq.heappop(h1))
            r.append((-h1[0] + h2[0]) / 2.0 if len(h1) == len(h2) else h2[0])
        return r


if __name__ == "__main__":
    obj = Solution()
    arr = [5, 15, 1, 3, 2, 8]
    medians = obj.getMedian(arr)
    print(" ".join(map(str, medians)))

    arr = [2, 2, 2, 2]
    medians = obj.getMedian(arr)
    print(" ".join(map(str, medians)))

File: 2025/02 February/20 Find median in a stream/main.cpp
// Link: https://www.geeksforgeeks.org/problems/find-median-in-a-stream-1587115620/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<double> getMedian(vector<int> &arr)
    {
        vector<int> sortedList;
        vector<double> medians;

        for (int num : arr)
        {
            auto it = lower_bound(sortedList.begin(), sortedList.end(), num);
            sortedList.insert(it, num);
            int n = sortedList.size();
            if (n % 2 == 1)
            {
                medians.push_back(sortedList[n / 2]);
            }
            else
            {
                medians.push_back((sortedList[n / 2 - 1] + sortedList[n / 2]) / 2.0);
            }
        }

        return medians;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {5, 15, 1, 3, 2, 8};
    vector<double> medians = object.getMedian(arr);
    for (double median : medians)
    {
        cout << median << " ";
    }
    cout << endl;

    arr = {2, 2, 2, 2};
    medians = object.getMedian(arr);
    for (double median : medians)
    {
        cout << median << " ";
    }
    cout << endl;
    return 0;
}